### PR TITLE
perf: optimize URL serialization

### PR DESCRIPTION
**Pull request by littledivy**
2.4x improvement in `(new URL(path)).pathname`.

```
# main
deno run -A --unstable cli/bench/url_parse.js 3 1000000
time 1080 ms rate 925925
time 1064 ms rate 939849
time 1048 ms rate 954198

# this patch
target/release/deno run -A --unstable cli/bench/url_parse.js 3 1000000
time 444 ms rate 2252252
time 429 ms rate 2331002
time 474 ms rate 2109704
```

### CODE REVIEW

File: ext/url/00_url.js [hunk 1]
```diff
@@ -444,20 +476,59 @@
...
       webidl.assertBranded(this, URLPrototype);
-      return this[_url].origin;
+      // https://github.com/servo/rust-url/blob/1d307ae51a28fecc630ecec03380788bfb03a643/url/src/origin.rs#L14
+      const scheme = this.#serialization.slice(0, this.#schemeEnd);
+      if (
+        scheme === "http" || scheme === "https" || scheme === "ftp" ||
+        scheme === "ws" || scheme === "wss"
+      ) {
+        if (this.port) {
+          return `${scheme}://${this.host}/${this.port}`;
```
Comment (aapoalas) [inline ext/url/00_url.js:494]: Port separated by `/`

**Review by littledivy — COMMENTED** [review]
Blocked on https://github.com/servo/rust-url/pull/788

File: ext/url/00_url.js [hunk 2]
```diff
@@ -363,10 +399,18 @@
...
+    }
+
     /** @return {string} */
     get hash() {
       webidl.assertBranded(this, URLPrototype);
-      return this[_url].hash;
+      // https://github.com/servo/rust-url/blob/1d307ae51a28fecc630ecec03380788bfb03a643/url/src/quirks.rs#L263
+      return this.#fragmentStart
+        ? trim(this.#serialization.slice(this.#fragmentStart))
+        : "";
```
Comment (lucacasonato) [inline ext/url/00_url.js:413]: I wonder if we should lazy cache this (applies to all part getters).

**Review by lucacasonato — COMMENTED** [review]
Looks great!

File: ext/url/00_url.js (hunk 2 above)
Comment (littledivy) [reply ext/url/00_url.js:413]: sure we can, not this PR though...already too much complexity - I also think we should do that global LRU cache first.

File: ext/url/lib.rs [hunk 3]
```diff
@@ -35,41 +37,71 @@ pub fn init() -> Extension {
...
-//   port: String,
-//   protocol: String,
-//   search: String,
-//   username: String,
-// }
-// TODO: implement cleaner & faster serialization
-type UrlParts = String;
-
 /// Parse `UrlParseArgs::href` with an optional `UrlParseArgs::base_href`, or an
 /// optional part to "set" after parsing. Return `UrlParts`.
```
Comment (lucacasonato) [inline ext/url/lib.rs:41]: Incorrect comment.

**Review by lucacasonato — APPROVED** [review]
Can you add a comment explaining how `op_url_parse`, `op_url_get_serialization`, and `op_url_parse_with_base` interact? Would be very helpful for future readers. After that, LGTM!

File: ext/url/00_url.js [hunk 4]
```diff
@@ -44,41 +44,37 @@
...
+    if (status === 0) {
+      return href;
+    } else if (status === 1) {
+      return core.ops.op_url_get_serialization();
+    } else {
+      throw new TypeError("Invalid URL");
+    }
   }
 
   class URLSearchParams {
```
Comment (billywhizz) [inline ext/url/00_url.js:80]: i did some benchmarking and there is a small change we can make here that passes all the wpt tests and bumps perf by 15-20% when api consumer accesses searchParams getter on URL and there are no query params, which is likely the common case. i can do this in a separate PR if you like as i think there are a few other small wins we can find here.

Screenshot from 2022-09-09 23-31-31

Comment (billywhizz) [reply ext/url/00_url.js:80]: actually, we could just move that up to the top of the function i think?

Comment (littledivy) [reply ext/url/00_url.js:80]: Nice find! 

> i can do this in a separate PR 

yup, that'll be great

**Pull request merged**